### ISSUE TITLE
Add touch support (BS $1000)

### ISSUE THREAD

**Issue by saltyfoam** [artf/grapesjs#241]
I have GrapesJS integrated into a solution being built and I'm looking for a developer to add touch support to GrapesJS, for our use and also to have it put out to the community. I am also looking for a developer to assist with GrapesJS integration and improvements for our site.
[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=48502777)](https://www.bountysource.com/issues/48502777-add-touch-support?utm_source=48502777&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by denim2x** [comment]
@saltyfoam Do you require *touch support* within the *editor* itself or within the *template*?

**Comment by saltyfoam** [comment]
I'd like to be able to use this editor on a touch device such as a tablet, so we need anything you can do with a mouse to be achievable with touch. I am currently using DragDropTouch.js for other parts of the Web site so if you could use that here it would be great as it seems to work for all the environments I want to support.

**Comment by saltyfoam** [comment]
How is this implementation coming along?

**Comment by cmcintosh** [comment]
If no-body has looked at taking this up, I may take a look at it.

**Comment by artf** [comment]
https://www.bountysource.com/issues/48502777-add-touch-support-bs-1000
seems like @denim2x still working on this

**Comment by saltyfoam** [comment]
Seems like there's been no activity on this for a month. I think cmcintosh could work on it.

**Comment by bayareawebpro** [comment]
Take a look at https://hammerjs.github.io

**Comment by Unsigno** [comment]
If you still interested on this feature and no one is working on it I would like to have a look .

I understand we are talking about handle user input on touch devices, no about responsive design ( can be improved but is not the target here ) . If some one can clarify i will be happy to start it :)

**Comment by saltyfoam** [comment]
This feature is still very important to my project and there is no one actively working on it. I believe it would also be great to have this feature for other people to use GrapesJS in a touch environment. At this point the goal is just to be able to create and edit a document with GrapesJS using a large touch enabled device such as a laptop, we can deal with smaller devices as a separate issue.

**Comment by Unsigno** [comment]
Oki, I'll check the code and if I think I can complete the issue I'll mark it as started. This should be forked from master and pushed to dev, right ?

**Comment by artf** [comment]
Hi @Unsigno you should only interact with the dev branch

**Comment by Unsigno** [comment]
It's being harder than I expected, but after some problems mapping the current code and looking for solutions, I think I'm on track to solve the dragging problems . I am basing my solution on the pruposed DragDropTouch polyfill, but it seems not supported on IE11 and Edge ( browsers using poiter events instead touch events ) is this acceptable ?

Are there another missing features for tocuh devices ? I was looking for it, but since i am not a real user and i can't test with all browsers,  I'd appreciate some reports.

**Comment by saltyfoam** [comment]
Hi,

We do support Edge on our Web site for drag and drop and touch and so I took a closer look and it looks like we’re also loading the following so perhaps with both polyfills we can get this to work on those environments.

https://github.com/CamHenlin/TouchPolyfill

Dominique

________________________________

**Comment by Unsigno** [comment]
Thanks, will take a look, i hope no issues overlaping polyfills. If it throw problems will implement the drag polyfill over pointers . 

Are there another missed features for touch devices ?

**Comment by saltyfoam** [comment]
We haven’t been able to use this framework at all on touch devices so I don’t know if there are other missing features.

We are currently loading both polyfills and using drag and drop so hopefully you won’t have conflicts either.

Dominique

On Mar 12, 2018, at 5:58 PM, Unsigno <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks, will take a look, i hope no issues overlaping polyfills. If it throw problems will implement the drag polyfill over pointers .

Are there another missed features for touch devices ?

**Comment by devontrae** [comment]
I think this would be pretty useful, so I'll take a crack at it as well.

**Comment by Unsigno** [comment]
@devontrae I am currently working on this , please look for another issue .

**Comment by devontrae** [comment]
@unsigno alright, well if you need a partner you know where to find me!

**Comment by UnitedPeace** [comment]
Have you had any luck with this issue?

**Comment by Unsigno** [comment]
No luck needed to solve issues :p 

Yes, there are some obstacles , but i still working on it. I have some things working but still some issues to fix, i hope send a PR soon .

**Comment by Unsigno** [comment]
I published a working version , still with some bugs , buts looks working and so you can test and report . https://github.com/Unsigno/grapesjs

Currently i am looking to fix the ColorPicker and trying to improve the DragDropTouch behavior .

@saltyfoam I added the touch polyfill and tested on IE/Edge ( on local version ) , it produces the same result on mouse and touch devices , but currently GrapesJS is not supported on IE/Edge . Since this should be a different issue , can i assume that producing the same result on touch and mouse as added touch support on that browsers ??

**Comment by saltyfoam** [comment]
Hi Unsigno,

I appreciate the work you've put into this! I tried out your solution with our existing code base on a Windows 10 laptop and I was able to drag components using touch but they would not drop. I tried on Edge, Chrome and Firefox with the same results.

I put together a package that reproduces our deployment so you could have a look. You can download it from http://uberfine.com/Test/GrapesJS3.zip and place it where your Web server can serve it, then load the Designer.htm into your Web browser.

At this point all I tried to do was drag the "Basic Header" component onto the top of the page. It works with the mouse but not with touch. The problem may be due to a difference between the JavaScript files I'm loading (such as DragDropTouch.js and touchpolyfill.min.js) and the ones you're using. This deployment was implemented by another engineer and he's no longer on the team so unfortunately I don't know much about how he integrated this.

Dominique

**Comment by saltyfoam** [comment]
I just tried the index.html in the grapesjs root folder and I am able to drop the “Block” element on Chrome using touch! So it would appear that the “drop” problem has something to do with my deployment, and not the environment.

Now just to add another wrinkle into all this, when I just retested with Edge I am able to not only drag but also drop using Edge! I guess I didn’t try it too much because I expected it to fail. I am not able to drop on Chrome or Firefox however.

So to recap, in my deployment (i.e. the package I provided) I can drag and drop on Edge using touch, and I can only drag on Chrome and Firefox using touch (no drop on these). Using index.html I am able to drag and drop in all three browsers.

Dominique

________________________________

**Comment by Unsigno** [comment]
Ok , i have some ideas to test, lets me some time to test and find a generic solution .

You could update DragDropTouch on your development too , because many changes are there and just the first "instance" apply .

**Comment by saltyfoam** [comment]
I updated my DragDropTouch.js from https://github.com/Bernardo-Castilho/dragdroptouch and it seems to behave the same way.

Dominique

________________________________

**Comment by Unsigno** [comment]
This is the older version , you can use https://github.com/Unsigno/grapesjs/blob/dev/src/utils/DragDropTouch.js . I don't created a new fork for that , but will do it .

**Comment by saltyfoam** [comment]
This version does resolve the drop issue on those two browsers!

I did some additional testing on other devices and I see one issue remaining. It affects the Chrome browser (iOS, Windows, Chromebook, Android, and iOS), the iOS native browser and Firefox (Windows) but strangely enough not Edge.

The issue is that when you start dragging an item from the right side panel (such as the Basic Header) the icon that is being dragged stops moving as soon as the green line appears in the main panel indicating where the object will be dropped. I can move my finger around and as a result move the green line around indicating that the drag operation is succeeding, but the icon does not follow my finger while it is over the white areas of the screen but is instead “stuck” to the right panel or the panel at the top (i.e. the black background areas). I’m not sure whether the problem is that the showing of the green line locks the icon in place or that the icon can’t escape the black areas of the screen (probably the latter?).

Dominique

________________________________

**Comment by Unsigno** [comment]
I'm glad to hear that. 
I know what you mean , its the latter one , it can't enter the iframe . I was looking on the colorpicker, and will look it tomorrow . I hope push this fixes tomorrow .

**Comment by Unsigno** [comment]
Hi , I updated my Grapes repo , it adds : 

- Touch events polyfill for IE/Edge
- Fix ColorPicker drag on touch devices 
- Fix DragDropTouch draggable thumbnail

* note, it updates DragDropTouch , so if you are using it on your main project you require overwrite that file 

Currently all looks working for me , wait for your new :)

**Comment by saltyfoam** [comment]
Hi,

I updated GrapesJS and DragDropTouch and things are working much better now. I did notice issues with the color picker. On IE it often doesn’t stay open and instead closes immediately. It also seems to result in the color #NaNNaNNaN in the edit box instead of the correct values for what I chose. I only did a bit of testing but this looks promising. 😊 I’ll do some more testing later today and if all is well I’ll look at how I release the funds to you.

On a different topic, I have a question for you. This work that you’ve been doing is in order to incorporate the GrapesJS into a product I’m building (see it at http://socialattache.com). GrapesJS is a small feature of the product - basically a Web Page / Landing Page / Email editor, but we have much more going on that will require a front end expert in JavaScript and CSS. Other libraries we use include jsPlumb and ckeditor and we’re also researching how to incorporate video and webinars. Would you be interested in joining the team, initially to complete the GrapesJS work and in future to work on other features? Our team members work at their own pace, as they have time.

Dominique

________________________________

**Comment by Unsigno** [comment]
Checking the issue.

Sounds good, you can send me a contat to gddg1989@gmail.com.

**Comment by Unsigno** [comment]
Ok , sorry for the timming , but i was losed with a error produced on IE emulator . After solving that, all works , and i can't reproduce your errors .

Are you testing on emulator or a real device ?? Is a single Grapes who produces that ??

In a single instance the colorpicker works for me , and i can't run your enviroment since i need some more files , and fixes .

You found some other bugs related to the issue , or i can send the PR to claim funds ??

**Comment by saltyfoam** [comment]
I will take another look later today and get back to you.

Dominique

On Apr 11, 2018, at 6:01 PM, Unsigno <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok , sorry for the timming , but i was losed with a error produced on IE emulator . After solving that, all works , and i can't reproduce your errors .

Are you testing on emulator or a real device ?? Is a single Grapes who produces that ??

In a single instance the colorpicker works for me , and i can't run your enviroment since i need some more files , and fixes .

You found some other bugs related to the issue , or i can send the PR to claim funds ??

**Comment by saltyfoam** [comment]
I am using real devices, no emulators.

What other files and fixes do you need to run  the package I sent you? The only bug I am aware of is the wrong case being used for WebSite.css at the bottom of Designer.htm (sorry about that).

I don’t know what you mean by « single Grapes ».

Dominique

On Apr 11, 2018, at 6:01 PM, Unsigno <notifications@github.com<mailto:notifications@github.com>> wrote:


Ok , sorry for the timming , but i was losed with a error produced on IE emulator . After solving that, all works , and i can't reproduce your errors .

Are you testing on emulator or a real device ?? Is a single Grapes who produces that ??

In a single instance the colorpicker works for me , and i can't run your enviroment since i need some more files , and fixes .

You found some other bugs related to the issue , or i can send the PR to claim funds ??

**Comment by saltyfoam** [comment]
Hi,

I put the package I sent you online and added your improvements so we can all test using the same code base. It’s available at:

http://uberfine.com/Test/GrapesJS3/Designer.htm

I recorded a video reproducing the two issues I see with the color picker:

http://uberfine.com/Test/GrapesJS3/ColorPickerBugs.wmv

When the color picker is working correctly I’ll release the funds to you as the rest seems to be working OK.

Dominique



________________________________

**Comment by Unsigno** [comment]
Oki , finally i fixed the tests and build the project . This should fix your issues . Try adding 'touchpolyfill' before 'DragDropTouch' too .

Wait for your news .

**Comment by saltyfoam** [comment]
If you like, feel free to send me a ZIP file of exactly what you want me to put up on my site and I’ll do that so we can both be testing with the same sources.

________________________________

**Comment by Unsigno** [comment]
Now you can update [grapes.min.js](https://github.com/Unsigno/grapesjs/blob/dev/dist/grapes.min.js) and [DragDropTouch.js](https://github.com/Unsigno/grapesjs/blob/dev/src/utils/DragDropTouch.js) files on your project from my repo, and in your Designer.html switch lines 17-18 just in case. I can't run your project on IE but if it works for you currently , this should solve the colorpicker issue.

If you have issues running your project on IE, you can clone my repo and run the example page.

Just tell me if you have some issue to run it or need some more help .

**Comment by Unsigno** [comment]
I have responded to your email and we can follow the thread there. 

About the video, I need to review the operation of the selector and the restrictions of the blocks to propose a solution, I know that there is already part of what you need defined, but I have to see how it works and its scalability. Will check your docs too and talk you later.

**Comment by tluyben** [comment]
Was this fix integrated in the main branch? The branch that @saltyfoam published on his site seems to work fine, but the main branch does not (I can drag & drop elements, but I cannot change their content on touch devices). 

Thanks!

**Comment by artf** [comment]
I didn't get further updates on the PR https://github.com/artf/grapesjs/pull/1042

**Comment by patricio-zenfolio** [comment]
Hi guys, not sure about the current status of this.
We are having issues on iOS touch devices:
- We included the plugin grapesjs-touch and drag&drop works fine.
- What is not working is the blocks selection and scrolling on the canvas section (nothing happens on touch)
- Kind of solved the first issue with the code hints from mikereem in https://github.com/artf/grapesjs-touch/issues/1

But the scrolling is not working.
Is working for anyone else?

On Android tablets all touch functionality seems to be working fine.
